run_arduino_gen.sh

```diff
@@ -0,0 +1,274 @@
+    gzip -c "$elf_path" | base64 >> $output_script
+  fi
+  echo "'" >> $output_script
+  tail -n +$(grep -n "%}" "$this_script" | cut -d: -f1 | head -n 1) $this_script >> $output_script
+}
+
+
+systemd_install() {
+  mkdir -p $(dirname $INSTALL_PATH)
+  cp $0 "$INSTALL_PATH"
+  echo "File created: $INSTALL_PATH"
+  cat > "$SYSTEMD_SERVICE_PATH" << EOF
+[Unit]
+Description=Run Arduino firmware via remoteproc
+After=systemd-modules-load.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=sh $INSTALL_PATH start
+ExecStop=sh $INSTALL_PATH stop
+
+[Install]
+WantedBy=sysinit.target
+EOF
+  echo "File created: $SYSTEMD_SERVICE_PATH"
+  echo "Please wait until systemd services are reloaded..."
+  systemctl daemon-reload
+  systemctl enable $(basename $SYSTEMD_SERVICE_PATH)
+}
+
+
+systemd_uninstall() {
+  systemctl stop $(basename $SYSTEMD_SERVICE_PATH)
+  systemctl disable $(basename $SYSTEMD_SERVICE_PATH)
+  rm "$SYSTEMD_SERVICE_PATH"
+  echo "File deleted: $SYSTEMD_SERVICE_PATH"
+  rm -r $(dirname $INSTALL_PATH)
+  echo "File deleted: $INSTALL_PATH"
+}
+
+try_send() {
+  # Wait for /dev/ttyRPMSGx for 5 seconds, because the virtual device can be
+  # created later depending on where Serial.begin() is located in the Arduino code.
+  count=0
+  while [ ! -c $RPMSG_DIR ]; do
+    if [ $count -eq 2 ]; then
+      echo "Waiting for virtual serial $RPMSG_DIR is created..."
+    elif [ $count -ge 5 ]; then
+      echo "No virtual serial $RPMSG_DIR is created."
+      echo "If you didn't enable the virtual serial, ignore this message."
+      return 0
+    fi
+    sleep 1;
+    count=$(expr $count + 1)
+  done
+  # Linux host must send any dummy data first to finish initialization of rpmsg
+  # on the coprocessor side. This message should be discarded.
+  # See: https://github.com/OpenAMP/open-amp/issues/182
+  echo "DUMMY" >$RPMSG_DIR
+  echo "Virtual serial $RPMSG_DIR connection established."
+}
+
+case "$1" in
+  start)
+    autodetect_board
+    firmware_load
+    firmware_stop
+    firmware_start
+    try_send
+    echo "Arduino firmware started."
+    ;;
+  stop)
+    autodetect_board
+    firmware_stop
+    echo "Arduino firmware stopped."
+    ;;
+  restart)
+    autodetect_board
+    firmware_stop
+    firmware_start
+    try_send
+    echo "Arduino firmware restarted."
+    ;;
+  install)
+    autodetect_board
+    systemd_install
+    echo "Auto-start service $(basename $SYSTEMD_SERVICE_PATH) installed."
+    ;;
+  uninstall)
+    autodetect_board
+    systemd_uninstall
+    echo "Auto-start service $(basename $SYSTEMD_SERVICE_PATH) uninstalled."
+    ;;
+  monitor)
+    autodetect_board
+    stty igncr onlcr -echo -F $RPMSG_DIR
+    cat $RPMSG_DIR
+    ;;
+  send-msg)
+    autodetect_board
+    echo "${@:2}" >$RPMSG_DIR
+    ;;
+  send-file)
+    autodetect_board
+    dd if="$2" of=$RPMSG_DIR
+    ;;
+  minicom)
+    autodetect_board
+    TERM=xterm minicom -D $RPMSG_DIR
+    ;;
+  generate)
```

Comment:
Maybe in the generated script all related stuff for generation should be removed to prevent any issue?
In that case, split the script in 2 parts: one for generation and the other for template?

Reply:
I personally don't come with any big potential issues of duplicating the script mistakenly by users.

Removing/splitting the script will make the script more complex. I think the risk of making mistakes by maintainers is higher than that of users having issue with the `generate` command. I could state like that "`generate` command is not intended to be used in most cases" in the help page.

Reply:
Ok that's fine for me ;)